tools/env-loader/pkg/envloader.go

```diff
@@ -69,35 +74,59 @@ func findGitRepoRoot() (string, error) {
 	}
 }
 
-// Finds environment value files for the given environment and value set, under the given directory.
-// File names will be returned in order of priority, with the lowest priority names first.
-func FindEnvironmentFilesInDirectory(environmentsDirectoryPath, environmentName, valueSet string) ([]string, error) {
-	commonFiles, err := filepath.Glob(filepath.Join(environmentsDirectoryPath, CommonFileGlob))
-	if err != nil {
-		return nil, trace.Wrap(err, "failed to find common value files")
-	}
+// Given a base path (typically the environments directory) and a relative subdirectory path
+// (typically an environment name), find all common value files in every directory between
+// the two, inclusively.
+// NOTE: this assumes '/' is used a directory separator character. This is important so that
+// the same results are produced on multiple platforms.
+func findCommonFilesInPath(basePath, relativeSubdirectoryPath string) ([]string, error) {
+	relativeSubdirectoryPath = filepath.Clean(filepath.ToSlash(relativeSubdirectoryPath))
+	subdirectoryNames := strings.Split(relativeSubdirectoryPath, EnvironmentNameDirectorySeparator)
+	directoryNamesToCheck := append([]string{"."}, subdirectoryNames...)
+
+	var commonFilePaths []string
+	currentDirectoryPath := basePath
+	for _, directoryNameToCheck := range directoryNamesToCheck {
+		currentDirectoryPath := filepath.Join(currentDirectoryPath, directoryNameToCheck)
+
+		fileInfo, err := os.Lstat(currentDirectoryPath)
```

Comment:
Could anything bad happen if this is a symlink? If an attacker can put a symlink at the right place can they make us write credentials somewhere we shouldn't?

Teleport has a number of "safe open" utilities after receiving several audit findings related to symlinks.

Reply:
> Could anything bad happen if this is a symlink? If an attacker can put a symlink at the right place can they make us write credentials somewhere we shouldn't?

Yes I _think_ so, in some corner cases. This is why I'm using `lstat` rather than `stat`, so that if the entry is a symlink the tool doesn't follow it.